In the Hacker News data set we'll compare these two types of posts to determine the following:
Do Ask HN or Show HN receive more comments on average?
Do posts created at a certain time receive more comments on average?

In [35]:
#REading the dataset
from csv import reader
open_file = open('hacker_news.csv')
read_file = reader(open_file)
hn = list(read_file)

In [36]:
hn[:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

In [37]:
headers = hn[0]

In [38]:
hn = hn[1:]

In [39]:
headers

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

In [40]:
hn[:5]

[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12']]

Extracting Ask HN and Show HN Posts
First, we'll identify posts that begin with either Ask HN or Show HN and separate the data for those two types of posts into different lists. Separating the data makes it easier to analyze in the following steps.

In [41]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(row)
    elif title.lower().startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)

print("The length of ask,show and other posts are:\n",len(ask_posts),len(show_posts),len(other_posts))    
        

The length of ask,show and other posts are:
 1744 1162 17194



Calculating the Average Number of Comments for Ask HN and Show HN Posts
Now that we separated ask posts and show posts into different lists, we'll calculate the average number of comments each type of post receives.

In [42]:
total_ask_comments = 0
for row in ask_posts:
    total_ask_comments += int(row[4])
avg_ask_comments = float(total_ask_comments)/float(len(ask_posts))
print(avg_ask_comments)

14.038417431192661


In [43]:
total_show_comments = 0
for row in show_posts:
    total_show_comments += int(row[4])
avg_show_comments = float(total_show_comments)/float(len(show_posts))
print(avg_show_comments)

10.31669535283993


On average, ask posts in our sample receive approximately 14 comments, whereas show posts receive approximately 10. Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

In [55]:
import datetime as dt
result_list = []

for row in ask_posts:
    comment = row[-1]
    num_comment = row[-3]
    comment_list = [comment , num_comment]
    result_list.append(comment_list)
    
counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    value = row[0]
    date,time = value.split()
    #check if same named variable can be used?
    time = dt.datetime.strptime(time , "%H:%M")
    #datetime.hour with str also used
    hour = time.strftime("%H")
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = int(row[1])
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += int(row[1])

comments_by_hour
counts_by_hour

{'00': 55,
 '01': 60,
 '02': 58,
 '03': 54,
 '04': 47,
 '05': 46,
 '06': 44,
 '07': 34,
 '08': 48,
 '09': 45,
 '10': 59,
 '11': 58,
 '12': 73,
 '13': 85,
 '14': 107,
 '15': 116,
 '16': 108,
 '17': 100,
 '18': 109,
 '19': 110,
 '20': 80,
 '21': 109,
 '22': 71,
 '23': 68}

In [59]:
avg_comments_by_hour = []

for key in comments_by_hour:
    value = comments_by_hour[key]/counts_by_hour[key]
    avg_comments_by_hour.append([key,value])
    
avg_comments_by_hour

[['18', 13.20183486238532],
 ['05', 10.08695652173913],
 ['16', 16.796296296296298],
 ['10', 13.440677966101696],
 ['15', 38.5948275862069],
 ['08', 10.25],
 ['14', 13.233644859813085],
 ['21', 16.009174311926607],
 ['03', 7.796296296296297],
 ['12', 9.41095890410959],
 ['06', 9.022727272727273],
 ['01', 11.383333333333333],
 ['00', 8.127272727272727],
 ['02', 23.810344827586206],
 ['17', 11.46],
 ['11', 11.051724137931034],
 ['04', 7.170212765957447],
 ['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['19', 10.8],
 ['23', 7.985294117647059],
 ['07', 7.852941176470588],
 ['20', 21.525],
 ['22', 6.746478873239437]]

In [78]:
swap_avg_by_hour = []
for row in avg_comments_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])

#print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour,reverse = True)
print("Top five hours for commenting:\n\n",sorted_swap[:5])

for value in sorted_swap[:5]:
    x = dt.datetime.strptime(value[1],"%H")
    x = x.strftime("%H:%M:")
    top_avg_comments = "{} {:.2f} average comments per post".format(x,value[0])
    print(top_avg_comments)




Top five hours for commenting:

 [[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21']]
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post



The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set documentation, the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.


Conclusion
In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.